accepted/file-enumeration.md

```diff
@@ -0,0 +1,385 @@
+    /// </summary>
+    public ref struct FindData<TState>
+    {
+        // This will have private members that hold the native data and
+        // will lazily fill in data for properties where such data is not
+        // immediately available in the current platform's native results.
+
+        // The full path to the directory the current result is in
+        public string Directory { get; }
+
+        // The full path to the starting directory for enumeration
+        public string OriginalDirectory { get; }
+
+        // The path to the starting directory as passed to the enumerable constructor
+        public string OriginalUserDirectory { get; }
+
+        // Note: using a span allows us to reduce unneeded allocations
+        public ReadOnlySpan<char> FileName { get; }
+        public FileAttributes Attributes { get; }
+        public long Length { get; }
+
+        public DateTime CreationTimeUtc { get; }
+        public DateTime LastAccessTimeUtc { get; }
+        public DateTime LastWriteTimeUtc { get; }
+
+        public TState State;
+    }
+}
+```
+
+### Transforms & Predicates
+
+We'll provide common predicates transforms for building searches.
+
+``` C#
+namespace System.IO
+{
+    internal static partial class FindPredicates
+    {
+        internal static bool NotDotOrDotDot(ref FindData findData)
+        internal static bool IsDirectory(ref FindData findData)
+    }
+
+    public static partial class FindTransforms
+    {
+        public static DirectoryInfo AsDirectoryInfo(ref FindData findData)
+        public static FileInfo AsFileInfo(ref FindData findData)
+        public static FileSystemInfo AsFileSystemInfo(ref FindData findData)
+        public static string AsFileName(ref FindData findData)
+        public static string AsFullPath(ref FindData findData)
+    }
+}
+
+```
+
+### DosMatcher
+
+We currently have an implementation of the algorithm used for matching files on Windows in FileSystemWatcher. Providing this publicly will allow consistently matching names cross platform according to Windows rules if such behavior is desired.
+
+``` C#
+namespace System.IO
+{
+    public static class DosMatcher
+    {
+        /// <summary>
+        /// Change '*' and '?' to '&lt;', '&gt;' and '"' to match Win32 behavior. For compatibility, Windows
+        /// changes some wildcards to provide a closer match to historical DOS 8.3 filename matching.
+        /// </summary>
+        public unsafe static string TranslateExpression(string expression)
+
+        /// <summary>
+        /// Return true if the given expression matches the given name.
+        /// </summary>
+        public unsafe static bool MatchPattern(string expression, ReadOnlySpan<char> name, bool ignoreCase = true)
+    }
+}
+```
+
+### Potential testability variant
+
+Having `FindData` be a struct is critical to meet the performance goals for the feature. To allow testing of Filters and Predicates we could also expose an interface of `IFindData`:
```

Comment:
We should probably discuss in the API review in more detail, but I feel  I'm not sure this minor tweak is worth the complexity. It seems in order to support testing, we'd need to allow passing something around that can serve to *produce* `FindData<T>` instances. In other words, for testing it seems just being able to to query a single find data is very limited, compared to an `IFileSystem` (whatever that would look like).